src/storage/src/hummock/sstable/builder.rs

```diff
@@ -235,21 +229,68 @@ impl<W: SstableWriter, F: FilterBuilder> SstableBuilder<W, F> {
         &mut self,
         full_key: FullKey<&[u8]>,
         value: HummockValue<&[u8]>,
-        is_new_user_key: bool,
     ) -> HummockResult<()> {
-        self.add(full_key, value, is_new_user_key).await
+        self.add(full_key, value).await
+    }
+
+    /// Add raw data of block to sstable. return false means fallback
+    pub async fn add_raw_block(
+        &mut self,
+        buf: Bytes,
+        filter_data: Vec<u8>,
+        smallest_key: FullKey<Vec<u8>>,
+        largest_key: Vec<u8>,
+        mut meta: BlockMeta,
+    ) -> HummockResult<bool> {
+        let table_id = smallest_key.user_key.table_id.table_id;
+        if self.last_table_id.is_none() || self.last_table_id.unwrap() != table_id {
+            self.table_ids.insert(table_id);
+            self.finalize_last_table_stats();
+            self.last_table_id = Some(table_id);
+        }
+        if !self.block_builder.is_empty() {
+            let min_block_size = std::cmp::min(MIN_BLOCK_SIZE, self.options.block_capacity / 4);
+            if self.block_builder.approximate_len() < min_block_size {
+                let block = Block::decode(buf, meta.uncompressed_size as usize)?;
+                let mut iter = BlockIterator::new(BlockHolder::from_owned_block(Box::new(block)));
+                iter.seek_to_first();
+                while iter.is_valid() {
+                    let value = HummockValue::from_slice(iter.value())
+                        .expect("decode failed for fast compact");
+                    self.add_impl(iter.key(), value, false).await?;
```

Comment:
I think it is possible to mix data from two tables into one block. If it is unlikely to create very small blocks, how about keeping the original implementation and optimize it when we found that small block is indeed an issue?

Reply:
This algorithm only works when all input sstable-files only include one state-table. So this condition will never happen.